src/main/java/org/opensearch/knn/index/query/KNNQueryBuilder.java

```diff
@@ -662,9 +662,24 @@ public String getWriteableName() {
 
     @Override
     protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) throws IOException {
-        // rewrite filter query if it exists to avoid runtime errors in next steps of query phase
+        QueryBuilder rewrittenFilter;
         if (Objects.nonNull(filter)) {
-            filter = filter.rewrite(queryShardContext);
+            rewrittenFilter = filter.rewrite(queryShardContext);
+            if (rewrittenFilter != filter) {
+                KNNQueryBuilder newKNNQuery = new KNNQueryBuilder(
```

Comment:
```suggestion
                KNNQueryBuilder rewrittenQueryBuilder = new KNNQueryBuilder(
```


nit: Another thing would be to use builder instead of constructor - KNNQueryBuilder.builder does validations on field level